### PR TITLE
Support `KerasLayer` model saving

### PR DESCRIPTION
**Pull request by eddddddy**
**Context:**
Part of the enhancements to the `qml.qnn` module.

**Description of the Change:**
Allow `KerasLayer` objects to be saved. We support the two main APIs: 
1. Model weights can be saved using `tf.keras.Model.save_weights`. The weights can be loaded again using `tf.keras.Model.load_weights`.
2. The entire model can be saved using `tf.keras.Model.save`. The model can be loaded again using `tf.keras.Model.load_model`.

### CODE REVIEW

File: pennylane/math/quantum.py [hunk 1]
```diff
@@ -429,6 +429,10 @@ def reduce_statevector(state, indices, check_state=False, c_dtype="complex128"):
     else:
         batch_dim, dim = np.shape(state)[:2]
 
+        # batch dim exists but is unknown; cast to int so that reshaping works
+        if batch_dim is None:
```
Comment (rmoyard) [inline pennylane/math/quantum.py:433]: Is it specific for tf autograph?

File: pennylane/qnn/keras.py [hunk 2]
```diff
@@ -347,7 +396,8 @@ def call(self, inputs):
 
         # reshape to the correct number of batch dims
         if has_batch_dim:
-            results = tf.reshape(results, (*batch_dims, *results.shape[1:]))
+            new_shape = tf.concat([batch_dims, tf.shape(results)[1:]], axis=0)
```
Comment (rmoyard) [inline pennylane/qnn/keras.py:399]: Why is this change necessary? Is saving impossible without?

File: tests/qnn/conftest.py [hunk 3]
```diff
@@ -44,9 +44,9 @@ def circuit(inputs, w1, w2, w3, w4, w5, w6, w7):
         qml.templates.StronglyEntanglingLayers(w1, wires=list(range(n_qubits)))
         qml.RX(w2[0], wires=0 % n_qubits)
         qml.RX(w3, wires=1 % n_qubits)
-        qml.Rot(*w4, wires=2 % n_qubits)
+        qml.Rot(w4[0], w4[1], w4[2], wires=2 % n_qubits)
```
Comment (rmoyard) [inline tests/qnn/conftest.py:47]: Is it not supported anymore?

File: pennylane/math/quantum.py (hunk 1 above)
Comment (eddddddy) [reply pennylane/math/quantum.py:433]: Yes, because during autograph compilation the batch dimension is unknown, e.g. the tensor has shape `(None, 2**n)` here.

File: tests/qnn/conftest.py (hunk 3 above)
Comment (eddddddy) [reply tests/qnn/conftest.py:47]: It's because Tensorflow doesn't allow iteration over abstract tensors. The old code only works when the `KerasLayer` is not wrapped with `tf.function`, but the new code works for both.

File: pennylane/qnn/keras.py (hunk 2 above)
Comment (eddddddy) [reply pennylane/qnn/keras.py:399]: It's necessary because the shapes need to be Tensorflow tensors in order for `None` dimensions to be compiled properly. Using tuples for the shapes does not work.

**Review by rmoyard — APPROVED** [review]
Thanks for answering my questions :100:

File: pennylane/qnn/keras.py [hunk 4]
```diff
@@ -169,6 +169,51 @@ def qnode(inputs, weights):
...
+
+        .. code-block::
+
+            model = tf.keras.models.load_model(SAVE_PATH)
+
+        .. note::
+
+            Currently ``KerasLayer`` objects cannot be saved in the ``HDF5`` file format. In order
+            to save a model using the latter method above, the ``SavedModel`` file format (default
+            in TensorFlow 2.x) should be used.
```
[the remark this answers is not in the document]
Comment (eddddddy) [reply pennylane/qnn/keras.py:215]: Yep, it works. Actually the bulk of the changes in the PR was to support this.

File: tests/qnn/test_keras.py [hunk 5]
```diff
@@ -573,22 +573,69 @@ def test_model_gradients(
...
+        new_res = new_model(x)
+        assert tf.math.reduce_all(res == new_res)
+
+    # the test is slow since TensorFlow needs to compile the execution graph
+    # in order to save the model
+    @pytest.mark.slow
+    @pytest.mark.parametrize("n_qubits, output_dim", indices_up_to(2))
+    def test_save_whole_model(self, model, n_qubits, tmpdir):
+        """Test if the entire model can be successfully saved and reloaded
+        using the .save() method"""
```
Comment (eddddddy) [inline tests/qnn/test_keras.py:621]: @trbromley

**Comment by eddddddy** [comment]
[sc-36428]

**Review by mudit2812 — APPROVED** [review]
Thanks @eddddddy . This is one of the few introductions to the qnn module I've had. The changes look good to me, and the testing is pretty rigorous, so I'm happy to approve.
